ffi/src/lib.rs

```diff
@@ -130,7 +186,16 @@ pub unsafe extern "C" fn fwd_root_hash(db: *mut Db) -> Value {
 /// This function is not exposed to the C API.
 /// It returns the current hash of an already-fetched database handle
 fn hash(db: &Db) -> Value {
-    let root = db.root_hash_sync().unwrap_or_default().unwrap_or_default();
+    let root = match db.root_hash_sync() {
+        Ok(Some(root)) => root,
+        Ok(None) => {
+            return unsafe { make_err_return("couldn't find root hash for db") };
+        }
+        Err(e) => {
+            return unsafe { make_err_return(&format!("couldn't get root hash: {}", e)) };
+        }
+    };
+
     Value::from(root.as_slice())
```

Comment:
It's best to avoid using `return` in rust. You can just use the value of the last expression, in this case it's a match expression:

```suggestion
        Ok(Some(root)) => root.into(),
        Ok(None) => "couldn't find root hash for db".into(),
        // I probably would skip adding context here and just use e.into(),
        Err(e) => (&format!("couldn't get root hash: {}", e)).into(),
        }
    }

```

You might even be able to do better than this if you have the right conversions set up. This would need a conversion from Error to Value, and from &str to Value, as shown in the other comments.
```
db.root_hash_sync()?.unwrap_or_else(|| "no root hash".into())

Reply:
Correct, you cannot, but you can `From<&str>`. `str` by itself has no length. This pattern is the same for several other types in rust `dyn Error` is another example. When you create a reference to something like that, the length information is stored in the fat pointer.